lib/internal/modules/cjs/loader.js

```diff
@@ -879,6 +879,46 @@ function createRequire(filename) {
 
 Module.createRequire = createRequire;
 
+Module.containsModuleSyntax = (source) => {
+  // Detect whether input source code contains at least one `import` or `export`
+  // statement. This can be used by dependent utilities as a way of detecting ES
+  // module source code from Script/CommonJS source code. Since our detection is
+  // so simple, we can avoid needing to use Acorn for a full parse; we can
+  // detect import or export statements just from the tokens. Also as of this
+  // writing, Acorn doesn't support import() expressions as they are only Stage
+  // 3; yet Node already supports them.
+  const acorn = require('internal/deps/acorn/acorn/dist/acorn');
+  source = stripShebang(source);
```

Comment:
the correct thing to use now is `stripShebangOrBOM` if you know that you're parsing cjs. If you're parsing ESM, you should only use `stripShebang`. This obviously presents a problem, and I don't have a solution for it.

Reply:
`stripShebangOrBOM` was removed in https://github.com/nodejs/node/commit/702331be906fe58e0ef66c7b31c7d2aeb3af3421, because the V8 parser handles stripping shebangs/[hashbangs](https://github.com/tc39/proposal-hashbang) now. Only `stripBOM` remains.

This is a problem, as Acorn doesn’t strip hashbangs. If I remove `stripShebang` from my method, the following throws:

```js
containsModuleSyntax('#!/usr/bin/env node\nimport "./file.js"');
```

When it should return `true`. I could resurrect `stripShebangOrBOM` for this method, I suppose?

I guess this is the problem with relying on Acorn for parsing in certain places (like the REPL) and V8 everywhere else.

Reply:
Nevermind, Acorn has an `allowHashBang` option, and it’s supported by its tokenizer. All is well. I rebased from the latest `master` and added a test with a shebang line.

Reply:
@GeoffreyBooth can we maybe have a preflight checklist re making sure that acorn configs/plugins... etc remain effective for this feature?

I'm thinking that if it is used sparsely, we might want to sanely keep track of dependencies so that updates don't break features like this.